pkg/metricprovider/cron_metric_provider.go

```diff
@@ -0,0 +1,455 @@
+	for i := 0; i < workers; i++ {
+		go wait.Until(p.runWorker, time.Second, stopCh)
+	}
+
+	klog.Infof("CronMetricProvider started")
+	<-stopCh
+	klog.Infof("CronMetricProvider stopped")
+}
+
+func (p *CronMetricProvider) runWorker() {
+	for p.processNextWorkItem() {
+	}
+}
+
+func (p *CronMetricProvider) processNextWorkItem() bool {
+	obj, shutdown := p.workqueue.Get()
+
+	if shutdown {
+		klog.Info("workequeue shutdown")
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.workqueue.Done.
+	err := func(obj interface{}) error {
+		// We call Done here so the workqueue knows we have finished
+		// processing this item. We also must remember to call Forget if we
+		// do not want this work item being re-queued. For example, we do
+		// not call Forget if a transient error occurs, instead the item is
+		// put back on the workqueue and attempted again after a back-off
+		// period.
+		defer p.workqueue.Done(obj)
+		var key *workItem
+		var ok bool
+		if key, ok = obj.(*workItem); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			p.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected workItem in workqueue but got %#v", obj))
+			return nil
+		}
+		if err := p.reconcileEHPAWorkItem(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			p.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%v': %v, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		p.workqueue.Forget(obj)
+		klog.Infof("Successfully synced '%v'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (p *CronMetricProvider) handleDeleteObj(obj interface{}) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			return
+		}
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+	}
+	item := &workItem{obj: object, itemType: workTypeDeleted}
+	klog.V(6).Infof("item %v deleted", item)
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) handleAddObj(obj interface{}) {
+	item := &workItem{obj: obj, itemType: workTypeAdded}
+	klog.V(6).Infof("item %v added", item)
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) handleUpdateObj(oldObj, newObj interface{}) {
+	item := &workItem{obj: newObj, oldObj: oldObj, itemType: workTypeUpdated}
+	klog.V(6).Infof("item %v updated", item)
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) reconcileEHPAWorkItem(key interface{}) error {
```

Comment:
We don't recommend using workType to check which event type is. This is not guaranteed when event queue do full sync. Actual how to handle reconcile in controller is very familiar when we use controller-runtime and we don't check eventType is Add or Update or Delete in that.
Please refering to https://github.com/kubernetes/sample-controller/blob/master/controller.go#L243 to implement sync loop.